Updated to parameter optimized by an evolutionary algorithm

ChristophSchranz opened the pull request:
Hi Nallath,
I was motivated to further improve this plugin when I saw the huge amount of downloads! 200k omg I couldn't believe it! :ok_man: :smile: 
As I'm currently into evolutionary algorithms, I decreased the miss-classification rate of the Tweaker on 100 objects by optimizing its given parameter. Those parameter were initially set by careful testing back then.
Can you please test the updates and let me know if it works within Cura?

Best,
Chris

P.S.: Currently, there trains a rather long evolutionary algorithm that will yield even better results than this one. I want to update the parameter in another PR next weeks.

ChristophSchranz commented:
Ohh there is one question:
In the line
 https://github.com/nallath/CuraOrientationPlugin/blob/c224458d78dfaff72643b4ca0f78c0359eeb2e5c/CalculateOrientationJob.py#L20 
you are extracting a variable to the parameter `min_volume`. What mode are you exactly using? I've optimized the objects with  `min_volume=False`, but the updated parameter should work better too.

If the mode is always  `min_volume=True`, I could also start a training with this parameter. This parameter, by the way, weights overhangs at 25mm height similar, below that less and higher overhangs more. The performance decreases just a little bit.

nallath commented:
By default, the min_volume parameter is set to True, but it can be changed to False by the user. I figured that the min_volume is probably what a user would want since long supports should also get a penalty (since it implies that there is more support material, and that is something you would want to minimize).

Soo, should I check & merge these values or wait for the updated data?

ChristophSchranz commented:
I guess the parameters should be optimized with `min_volume=True`. I'll start the algorithm today again, this will take a few days.

ChristophSchranz commented:
I've added another commit:
From now on, the used parameter set is dependent from the selected mode, either minimize faces, or volumes (that is here a combination of both, that is plausible for me).

Please let me know if the changes work as expected!

ChristophSchranz commented:
I updated also the parameter now for surface minimization.
And I fixed an undeclared variable `progress_callback`, it should now work again; and better for both modes.
I hope you find some time to merge it!

ChristophSchranz commented:
There is also no image logo for  the plugin anymore. Why that?

nallath commented on MeshTweaker.py:
```diff
@@ -35,35 +70,51 @@ class Tweak:
+
+        for k, v in parameter.items():
+            # print(f"{k} = {v}")
+            exec(f"self.{k} = {v}")
```
This is really not the right way to do this (exec is dangerous, and the f"" notation isn't supported in python3.5 (which is what Cura runs on).

There are a few ways to do this a bit better though. The quickest one would be:
```python3
for k, v in parameter.items():
    setattr(self, k, v)
```

There is also an alternative way of doing it, by overloading the getattr of the object iself.

Perhaps the best way of doing it, would be to actually add them as attributes and let the values be defined by the dict. That also allows for easier documentation with the parameters itself.

ChristophSchranz replied:
Thanks!
I've made the required changes.
`setattr` is useful for this parameter extraction, because it is limited to the parameters only.

ChristophSchranz replied:
I hope it can be merged now!

ChristophSchranz commented:
I hope you have time the next days to update it!

nallath commented:
Yeah, this got a bit drowned in the sheer volume of notifications i get from github!

nallath commented:
I've published it to the marketplace. Now it's awaiting review from my colleagues.

ChristophSchranz commented:
Super, thanks!

> Yeah, this got a bit drowned in the sheer volume of notifications i get from github!

I understand that, do you have suggestions on how to manage this (shared) repository better? 

I'd like to further develop the module, like adding a symmetry heuristic and/or GPU processing. However, I can't test the plugin within Cura yet and it is still a hobby project..

nallath commented:
As long as there are is an interface, it's not that hard for me to glue it to Cura (because that's pretty much what the plugin does). So even if it was just a compiled program, I could set it up to work with Cura (take libSavitar for instance, where we use C++ classes that are hooked up to python). It could even work a bit like the engine (stand-alone application that gets called by the plugin to do the heavy lifting). 

GPU processing is a fickle beast in my experience, especially if you want to support multiple operating systems.

ChristophSchranz commented:
Ahh okay, that's good to know!

I've also seen that only you are the author of this plugin. Is it possible to list us both? Here, the core algorithm was developed and is maintained by me. Nonetheless, your feedback was very important for the performance of the current version!

nallath commented:
> Ahh okay, that's good to know!
> 
> I've also seen that only you are the author of this plugin. Is it possible to list us both? Here, the core algorithm was developed and is maintained by me. Nonetheless, your feedback was very important for the performance of the current version!

On the marketplace it's not possible, since it needs to be tied to a single account. As for the plugin information itself, It's a freeform text, so that should be possible. On the marketplace, I've also added your name as the author of the algorithm since the beginning.
